### PR TITLE
Convert #4 into a sorted list of things

### PR DESCRIPTION
**Pull request by astorije**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,14 +2,69 @@
...
 
+- Client plugin system
+  - Users can enable plugins from the app
+  - A plugin either gets downloaded on the server the first time it is enabled, or reused if others are already using it
+    - Major security flaw: create an evil plugin, have someone enable it, that's custom code running on the server...
+  - Client plugins update by themselves
+    - Every half-hour? More? Less? Different strategy?
+  - Which strategy to gather and list the plugins
+- Plugins
+  - GitHub Flavored Markdown support
```
Comment (JocelynDelalande) [inline README.md]: Not sure I'm in favor of that, the other clients don't support that, IRC is bare text. Although there are some conventions like `*important*` or `**more important**`. Anyway, the github flavor brings nothing to markdown-over-irc IMHO over standard¹ markdown.

¹ lol

File: README.md [hunk 2]
```diff
@@ -2,14 +2,69 @@
...
+- Functional testing for the back-end
+- UI testing for the front-end
+- Reach code coverage by layers, set up thresholds and due dates
+
 ## Code and project organization
+
+- `shout-irc` organization for the projects and the community
+- Continuous Integration
+  - Linux testing with Travis CI
+  - Windows testing with AppVeyor
```
Comment (JocelynDelalande) [inline README.md]: I'm only in favor of building and testing for windows if we do have at least a windows user among moderators team.

File: README.md [hunk 3]
```diff
@@ -2,14 +2,69 @@
...
+- Client plugin system
+  - Admin can blacklist client plugins
+
 ## Client ↔ Server Communication Protocol
 
+- Websocket, define exchange vocabulary
+
 ## Client features
 
+- Client plugin system
```
Comment (JocelynDelalande) [inline README.md]: I don't understand, you speak about client plugin system, but with code running on the server… Could you detail a bit what you are thinking about ?

**Comment by JocelynDelalande** [comment]
Mostly in favor of what you wrote.

Made me think about creating #7 on privacy

File: README.md (hunk 3 above)
Comment (astorije) [inline README.md]: So, there are two parts in Shout: the client and the server. As you can imagine, the client is the JS code running in the user's browser, while the the server is JS code running on the server, even when the user is not logged in.

Some plugins will only be for the client, while some others will only be for the server. They cannot be mixed with each other, and they will need to be located in 2 places.

Another way you can look at it: server plugins are enabled by the administrator who installs, configures, updates and runs Shout, while client plugins are enabled at the client level, by the users.

File: README.md (hunk 1 above)
Comment (astorije) [inline README.md]: So, first of all, whether you like it or not, GFM is becoming the de-facto standard :-) GitHub and *all* related tools (such as Gitter) are using GFM, as well as Stack Overflow websites, Jekyll, Atom, ...

But anyway, my point was not to choose one of the many self-called standard Markdown (yet another bikeshedding topic). I was just trying to remember what tickets were deemed to be plugin-worthy, and this was one: https://github.com/erming/shout/issues/435

Whether we are in favor or not is irrelevant: it's very unlikely any of us (the project maintainers) will come up with such plugin ourselves. However, listing was sorts of plugins are desired from the community will help us defining the boundaries of the plugin system.

Comment (JocelynDelalande) [inline README.md]: Ok, would reword into "plugin requests examples"

File: README.md (hunk 2 above)
Comment (Xe) [inline README.md]: Windows support is not worth your time.

Comment (astorije) [inline README.md]: @Xe, I'd strongly argue it is. Shout is aimed to be a simple tool for people not necessarily in tech. As Node.js is fully available for Windows and Shout can be run on a Windows desktop machine (I know a few people who run it just like this), it's free to let AppVeyor test/run the server. With the caveat that we do not fully support Windows as a server (and not focusing on fixing it either), people will contribute if they need.
However, as client should be tested/testable, it *must* run on Windows.

**Comment by astorije** [comment]
@JocelynDelalande, amended with your comments.

File: README.md [hunk 4]
```diff
@@ -2,14 +2,75 @@
...
+  - Users can enable plugins from the app
```
Comment (JocelynDelalande) [inline README.md]: Yes, but against the fact they get installed by the users, *enable/disable* from browser is fine to me, but not *installing* by (non-admin user) seems like security-flaw-prone, and a bit klunky, as administrator, you're never sure if you break something upgrading shout.

Install client plugins from browser is taking risks, too much promise and too much unreliability IMHO.

As shout targets firstly small-sized instances (or even medium), your admin is never far away to install a plugin, and a list of good-quality maintained plugins could be shipped with shout with some guarantee of consistency between shout version and plugins version (yeah, I know, well designed and tested API can survive internal changes, but that's the theory...).

I'm not even sure auto-update is a good idea neither. Installing a plugin one-time gives you opportunity for review (maybe from someone else, not from you), auto-update is trusting a code and all the future states of it, it's more risky, features/maintainers change across time.

Comment (astorije) [inline README.md]: First of, I never said my proposition was the best one, nor the only one ;-)
However, although security is an aspect that cannot be undetaken, I am keeping usability and user experience as a first-class requirement.

> Install client plugins from browser is taking risks

Don't forget that a client plugin here is front-end code that can only run in the client, and will be loaded from the HTML essentially. Risks are pretty inexistant, for the server I mean.

> As shout targets firstly small-sized instances (or even medium), your admin is never far away to install a plugin

Yikes, that sounds like the worst UX possible to deal with plugins. If I had to ask an administrator any time I want to play with a plugin on a webapp, I would live without plugins.

> I'm not even sure auto-update is a good idea neither.

In solutions where users deal with their plugins themselves, yes auto-update is the right way to go, otherwise only a fraction will update, and even those won't be as thorough as you suggest. In solutions where administrators are in charge of client plugins, then I agree it can be disabled by default.

Comment (JocelynDelalande) [inline README.md]: @astorije Ok, no consensus at the moment between the two of us. I have a suggesiton to go forward : let's not be accurate on the details of client-side plugins, and discuss that with more people later, what do you think ?

Comment (astorije) [inline README.md]: Oh yes, definitely, group discussions are mandatory.

Plus in the end we kind of bikeshedded here: we extensively debated on how to install and manage plugins while we haven't even started the discussion on what is the actual plugin mechanism! (API, ...)
My latest amend actually lists a third way of doing things: plugins within shout repository itself **in a first iteration**. What is really wanted right now is a way to develop features that are usable *and* maintainable, as we are already saying no to features as we'd prefer them as plugins.
If we can at least get to split the code in independent chunks, then that's already a win. How to externalize and manage them is only a second step.

So I'd say let's future ourselves deal with that, and focus on the actual plugin system discussion.

**Comment by JocelynDelalande** [comment]
Ok, that's quite verbose and could be refined for a roadmap, but heck, it's a tool, let's start with something :) :+1:

**Comment by astorije** [comment]
\o/
Exactly, we need to refine, discuss, improve, but now we have a base. Even
if no one agrees, it will create emotion and drive people to PR... Right?
:-)
